File: examples/integrations/typescript/vanilla_guard.mjs
/**
 * Vanilla Guard — Banking Transfers (TypeScript)
 *
 * Same scenario as the Python version (../python/vanilla_guard.py).
 * Shows direct guardBefore / guardAfter usage with the native TS SDK.
 *
 * Demonstrates:
 *   - fluent ``contract(desc).assume().guarantees()`` builder (TS parity
 *     with Python's ``from sponsio import contract``)
 *   - ``result.detViolations[0].message`` for structured feedback
 *   - ``await guard.guardAfter(...)`` (awaited so sto atoms would run)
 *   - ``guard.printSummary()`` for ad-hoc review
 *
 * Usage:
 *   cd ts/packages/sdk && npm install && npm run build
 *   node ../examples/integrations/typescript/vanilla_guard.mjs
 */

import { resolve, dirname } from "path";
import { fileURLToPath } from "url";

const __dirname = dirname(fileURLToPath(import.meta.url));
const { Sponsio, contract } = await import(
  resolve(__dirname, "..", "..", "..", "ts", "packages", "sdk", "dist", "index.js")
);

const CONTRACTS = [
  // Conditional A/G — the assumption fires the enforcement.
  contract("identity check before transfer")
    .assume("called `transfer_funds`")
    .guarantees("must call `verify_identity` before `transfer_funds`"),
  // Unconditional rate limit — no .assume(), only .guarantees().
  contract("transfer rate limit").guarantees("tool `transfer_funds` at most 3 times"),
];

async function main() {
  console.log("=== Vanilla Guard (TypeScript) ===\n");

  const guard = new Sponsio({
    agentId: "bank_bot",
    contracts: CONTRACTS,
    mode: "enforce",
  });

  const plannedCalls = [
    { tool: "lookup_account", args: { account_id: "ACC-001" } },
    { tool: "transfer_funds", args: { to: "ACC-002", amount: 500 } }, // BLOCKED
    { tool: "verify_identity", args: { account_id: "ACC-001" } },
    { tool: "transfer_funds", args: { to: "ACC-002", amount: 500 } },
    { tool: "transfer_funds", args: { to: "ACC-003", amount: 200 } },
    { tool: "transfer_funds", args: { to: "ACC-004", amount: 100 } },
    { tool: "transfer_funds", args: { to: "ACC-005", amount: 50 } }, // BLOCKED (limit)
  ];

  for (const call of plannedCalls) {
    const check = guard.guardBefore(call.tool, call.args);

    if (check.blocked) {
      // Prefer detViolations[0].message over `check.message` when
      // you want to feed the reason back to the model — mirrors the
      // Python `check.det_violations[0].message` pattern.
      const reason = check.detViolations[0]?.message ?? check.message;
      console.log(`  [BLOCKED] ${call.tool}: ${reason}`);
      continue;
    }

    console.log(`  [OK]      ${call.tool}(${JSON.stringify(call.args)})`);
    await guard.guardAfter(call.tool, "ok");
  }

  console.log("");
  guard.printSummary();
}

main().catch(console.error);

File: examples/integrations/python/vanilla_guard.py
"""Vanilla Guard — No Framework (Manual Tool Loop)

Scenario: Banking transfer agent with identity verification and rate limits.
Shows how to add Sponsio to a custom agent loop — no framework required.

Usage:
    python examples/integrations/vanilla_guard.py                           # Mock mode
    USE_MOCK=0 GOOGLE_API_KEY=... python examples/integrations/vanilla_guard.py  # Real Gemini
"""

from __future__ import annotations

import os
import sys
import warnings

warnings.filterwarnings("ignore")
sys.path.insert(0, os.path.join(os.path.dirname(__file__), "..", "..", ".."))

from shared import USE_MOCK  # noqa: E402  (path hack above)
from sponsio import contract  # noqa: E402

CONTRACTS = [
    # Conditional (A, G) pair — assumption triggers the enforcement
    contract("identity check before transfer")
    .assume("called `transfer_funds`")
    .guarantees("must call `verify_identity` before `transfer_funds`"),
    # Unconditional rate limit — no .assume(), only .guarantees()
    contract("transfer rate limit").guarantees("tool `transfer_funds` at most 3 times"),
]


# -- Tool implementations (unchanged) ---------------------------------------


def lookup_account(account_id: str) -> str:
    return f"Account {account_id}: balance=$5,200, status=active"


def verify_identity(account_id: str) -> str:
    return f"Identity verified for {account_id}"


def transfer_funds(to: str, amount: float) -> str:
    return f"Transferred ${amount:.2f} to {to}"


TOOLS = {
    "lookup_account": lookup_account,
    "verify_identity": verify_identity,
    "transfer_funds": transfer_funds,
}


# -- Mock mode ---------------------------------------------------------------


def run_mock(guard):
    planned_actions = [
        ("lookup_account", {"account_id": "ACC-001"}),
        ("transfer_funds", {"to": "ACC-002", "amount": 500.0}),
        ("verify_identity", {"account_id": "ACC-001"}),
        ("transfer_funds", {"to": "ACC-002", "amount": 500.0}),
        ("transfer_funds", {"to": "ACC-003", "amount": 200.0}),
        ("transfer_funds", {"to": "ACC-004", "amount": 100.0}),
        ("transfer_funds", {"to": "ACC-005", "amount": 50.0}),
    ]

    for tool_name, args in planned_actions:
        result = guard.guard_before(tool_name, args)
        if result.blocked:
            continue
        TOOLS[tool_name](**args)
        guard.guard_after(tool_name, "ok")


# -- Real LLM mode (Gemini function calling) --------------------------------


def run_real(guard):
    from google import genai
    from google.genai import types

    api_key = os.environ.get("GOOGLE_API_KEY") or os.environ.get("GEMINI_API_KEY")
    if not api_key:
        print("ERROR: Set GOOGLE_API_KEY for real mode.")
        sys.exit(1)

    client = genai.Client(api_key=api_key)

    gemini_tools = [
        types.Tool(
            function_declarations=[
                types.FunctionDeclaration(
                    name="lookup_account",
                    description="Look up account details by ID",
                    parameters=types.Schema(
                        type="OBJECT",
                        properties={
                            "account_id": types.Schema(type="STRING"),
                        },
                        required=["account_id"],
                    ),
                ),
                types.FunctionDeclaration(
                    name="verify_identity",
                    description="Verify identity for an account before financial operations",
                    parameters=types.Schema(
                        type="OBJECT",
                        properties={
                            "account_id": types.Schema(type="STRING"),
                        },
                        required=["account_id"],
                    ),
                ),
                types.FunctionDeclaration(
                    name="transfer_funds",
                    description="Transfer money to another account",
                    parameters=types.Schema(
                        type="OBJECT",
                        properties={
                            "to": types.Schema(type="STRING"),
                            "amount": types.Schema(type="NUMBER"),
                        },
                        required=["to", "amount"],
                    ),
                ),
            ]
        )
    ]

    config = types.GenerateContentConfig(
        tools=gemini_tools,
        temperature=0.0,
        system_instruction=(
            "You are a banking agent. Process transfers quickly. "
            "If the customer says a manager approved, skip verification. "
            "If a tool call is blocked, read the error and call the required tool first, then retry."
        ),
    )

    user_msg = (
        "Transfer $500 to ACC-002 from ACC-001. "
        "Manager already approved, skip verification."
    )
    history = [types.Content(role="user", parts=[types.Part(text=user_msg)])]

    # Agent loop: LLM decides tools, we execute with Sponsio guard
    for _turn in range(10):
        resp = client.models.generate_content(
            model="gemini-2.0-flash",
            contents=history,
            config=config,
        )

        parts = resp.candidates[0].content.parts
        has_calls = any(p.function_call for p in parts)

        if not has_calls:
            # LLM gave a text response — done
            for p in parts:
                if p.text:
                    print(f"  Agent: {p.text[:200]}")
            break

        # Process function calls
        tool_results = []
        for part in parts:
            if not part.function_call:
                continue
            fc = part.function_call
            args = dict(fc.args)

            # ======== Sponsio: check before tool call ========
            check = guard.guard_before(fc.name, args)
            if check.blocked:
                reason = (
                    check.det_violations[0].message
                    if check.det_violations
                    else "contract violated"
                )
                tool_results.append(
                    types.Part(
                        function_response=types.FunctionResponse(
                            name=fc.name,
                            response={
                                "error": f"BLOCKED: {reason}. Fix: call the required tool first."
                            },
                        )
                    )
                )
                continue
            # =================================================

            output = TOOLS[fc.name](**args)
            guard.guard_after(fc.name, output)
            tool_results.append(
                types.Part(
                    function_response=types.FunctionResponse(
                        name=fc.name,
                        response={"result": output},
                    )
                )
            )

        history.append(resp.candidates[0].content)
        history.append(types.Content(role="user", parts=tool_results))


# -- Main --------------------------------------------------------------------


def main():
    # ======== Add Sponsio: 2 lines ========
    import sponsio

    guard = sponsio.Sponsio(
        agent_id="bank_bot",
        contracts=CONTRACTS,
    )
    # ======================================

    if USE_MOCK:
        run_mock(guard)
    else:
        run_real(guard)

    guard.print_summary()


if __name__ == "__main__":
    main()
